compiler/src/yul/runtime/functions/data.rs

```diff
@@ -380,3 +381,29 @@ pub fn load_data_string() -> yul::Statement {
         }
     }
 }
+
+/// Revert with encoded reason string
+pub fn revert_with_reason_string() -> yul::Statement {
+    function_definition! {
+        function revert_with_reason_string(reason) {
+            (let ptr := alloc(4))
+            // Function selector for Error(string)
+            (mstore(ptr, 3963877391197344453575983046348115674221700746820753546331534351508065746944))
+
+            // Write the (fixed) data offset into the next 32 bytes of memory
+            (let __ := alloc_mstoren(0x0000000000000000000000000000000000000000000000000000000000000020, 32))
```

Comment:
I have to confess that I don't fully grasp the rationale for this :point_up: I tried to follow the conversation inhttps://github.com/ethereum/EIPs/issues/838 but it doesn't go into much detail (or not enough for me to properly understand it). I *think* this *data offset* of `32` is just meant to say that the actual reason string is found at an offset of `32` bytes simply because thats how ABI encoded strings are defined. But to me it seems pointless to have this. After all `Error(string)` returns a `string` so it should be clear that there's a `32` data offset but maybe I'm just misunderstanding the whole thing haha. Anyway, this is how the encoding works and the tests prove that it is working in the same way as solidity.

Reply:
>It makes sense when there's more items being encoded.

Do you have an example where the `dataoffset` becomes useful? Let's say we have `Error(AnyOtherABIType)`. Why would I need the `dataoffset` for anything? Wouldn't it work just as fine without it? Whatever follows the *selector* would be the data that is decoded as the ABI type that is specified with the selector.

Reply:
If we're encoding something like `(string, string)`, we need to provide the offsets to both pieces of dynamically sized data. This way we only need to read the offset, size, and the data when reading each string. If offsets were not provided, we would need to read each dynamic size and sum them together to get the offset. This would make implementations much more error prone.

For reference, the encoding would look roughly like this:

`[selector], [offset_1, offset_2], [(size_1, data_1), (size_2, data_2)]` 

With a single dynamic element, it's not necessary to provide this offset, but it's best to make a few simple rules with encoding schemes and stick with them - even if it seems silly at times.

Reply:
>if offsets were not provided, we would need to read each dynamic size and sum them together to get the offset

But isn't that what I would also have to do if I'm reading the output of `pub foo -> (string, string)`? In that case, the returned data will also only be `[(size_1, data_1), (size_2, data_2)]` without providing me the offsets separately, no?

Reply:
The encoding of `(string, string)` would include the offsets of each string item. Then to decode, we read the offset, size, and data for each string.

Is there something I'm missing?

Reply:
The [examples](https://docs.soliditylang.org/en/v0.8.4/abi-spec.html#use-of-dynamic-types) are quite helpful.